test/tools/yulopti.cpp

```diff
@@ -76,58 +76,66 @@ namespace po = boost::program_options;
 class YulOpti
 {
 public:
-	void printErrors()
+	static void printErrors(CharStream const& _charStream, ErrorList const& _errors)
 	{
 		SourceReferenceFormatter{
 			cerr,
-			SingletonCharStreamProvider(*m_charStream),
+			SingletonCharStreamProvider(_charStream),
 			true,
 			false
-		}.printErrorInformation(m_errors);
+		}.printErrorInformation(_errors);
 	}
 
-	bool parse(string const& _input)
+	void parse(string const& _input)
 	{
-		ErrorReporter errorReporter(m_errors);
-		m_charStream = make_shared<CharStream>(_input, "");
-		m_ast = yul::Parser(errorReporter, m_dialect).parse(*m_charStream);
-		if (!m_ast || !errorReporter.errors().empty())
+		ErrorList errors;
+		ErrorReporter errorReporter(errors);
+		CharStream _charStream(_input, "");
+		try
 		{
-			cerr << "Error parsing source." << endl;
-			printErrors();
-			return false;
+			m_ast = yul::Parser(errorReporter, m_dialect).parse(_charStream);
+			if (!m_ast || !errorReporter.errors().empty())
+			{
+				cerr << "Error parsing source." << endl;
+				printErrors(_charStream, errors);
+				throw std::runtime_error("Could not parse source.");
+			}
+			m_analysisInfo = make_unique<yul::AsmAnalysisInfo>();
+			AsmAnalyzer analyzer(
+				*m_analysisInfo,
+				errorReporter,
+				m_dialect
+			);
+			if (!analyzer.analyze(*m_ast) || !errorReporter.errors().empty())
+			{
+				cerr << "Error analyzing source." << endl;
+				printErrors(_charStream, errors);
+				throw std::runtime_error("Could not analyze source.");
```

Comment:
Why not rethrow the original?

```suggestion
				throw;
```

Reply:
Ah, disregard this :) It's not inside the `catch` block.

Reply:
It's not an exception handler - there is no original exception here :-).

Reply:
In the ``catch(...)`` block below I do rethrow, but here I have nothing to throw, so I make up something new to throw :-).
I guess throwing nothing would also be fine, though, since the error was already printed and we have a catch-all handler in main...

Reply:
Actually, I had to check what happens if you ``throw;`` without active exception :-). Could have imagined both: termination and something like throwing void - but it in fact aborts, which makes more sense - so I need to throw *something* anyways, and then a ``runtime_error`` is as good as anything I guess.

Reply:
> Ah, disregard this :) It's not inside the `catch` block.

Only seeing that now, but yeah :-).

Reply:
Right, sorry. I commented too fast, before realizing that I'm not looking at the `catch` block. I edited my comment but you also responded pretty fast so you probably caught the original comment :)